### PR TITLE
feat: clickable link, underline and strike-through

### LINKED ISSUE #758: feat: ability to have a cell be a clickable link
Idk how we'd want to do this from a data perspective — like a cell could be a string with an optional type of "link" — but from a user's perspective it'd be great if text formatted as URLs became interactive and you could click on them.

### VSCode as an example

VSCode is a good example of implementation, I think, because the ability for something to be clicked as a link is _secondary_. It does not take over the primary function of the app, which is editing text.

For example, if you come across a URL in code, it is underlined to indicate it's a link and it can be interacted with:

CleanShot 2023-10-30 at 11 43 30@2x

But clicking on it does the same thing that clicking on any other text does: it positions the cursor. Just because it's a link doesn't mean it overrides the default functionality of clicking in text files.

However, if you hover the text you get a little hint telling you how to interact with it.

CleanShot 2023-10-30 at 11 43 35@2x

Then you learn to click on links in VSCode you hit CMD + click and anything that's underlined as a URL can be opened in a browser.

### In Quadratic

Something similar in quadratic would be nice. Maybe a good place to start would be we detect if something's a link rather than giving the ability to users to define cells as links. That might come later?

As an example, [I created this sheet](https://app.quadratichq.com/file/348ffe6d-4e95-434b-b850-392ab121b5af) where you can input a web page's address and it'll parse out all the URLs on that page and enumerate them.

CleanShot 2023-10-30 at 11 41 55@2x

It would be nice if I could click on any of those and open them in a browser.

### PR DESCRIPTION
**Pull request by AyushAgrawal-A2**
closes #758

### CODE REVIEW

**Comment by luke-quadratic** [comment]
Doesn't seem to be working 
CleanShot 2024-09-03 at 08 46 59

**Comment by luke-quadratic** [comment]
@jimniels @davidkircos @AyushAgrawal-A2 These powder blues in the sheet are so light - wdyt about matching more with the direction colors have started to go in the app? The colors are now split and disjointed with blue and powder blue in the same experiences 

CleanShot 2024-09-03 at 08 48 57@2x

**Comment by AyushAgrawal-A2** [comment]
@luke-quadratic 
>Doesn't seem to be working
currently it is Cmd + click to open, cursor pointer pn hover is only meant to indicate it is clickable... open to changing this behaviour

**Comment by jimniels** [comment]
> @jimniels @davidkircos @AyushAgrawal-A2 These powder blues in the sheet are so light - wdyt about matching more with the direction colors have started to go in the app?

Agree that this blue we use is not a very high contrast color and surely fails accessibility.

FWIW: I'm planning on doing this in the "new app design work" where I want to make the cursor and other elements the same "blue" as we use elsewhere, e.g.

CleanShot 2024-09-03 at 09 04 38@2x

May as well start that work in this PR I would say.

**Review by davidkircos — CHANGES_REQUESTED** [review]
Links don't work for me

https://github.com/user-attachments/assets/138b88bf-5090-4c55-bc04-39c7ebbde96c

**Comment by luke-quadratic** [comment]
> Links don't work for me
> 
>  CleanShot.2024-09-11.at.14.42.34.mp4

@davidkircos see conversation above, cmd+click

**Review by luke-quadratic — APPROVED** [review]
works great on functionality

**Review by davidfig — APPROVED** [review]

**Comment by davidfig** [comment]
Looks much better

**Review by davidfig — APPROVED** [review]

**Comment by jimniels** [comment]
The link tooltip should have a delay. Right now it appears _immediately_ when you hover over it, so if you're moving your cursor across the sheet and go across a cell it's going to show even if you didn't mean to. It should require a few hundred millisecond hover intention to show it (this is how VSCode does it). 

If we don't have it, it gets especially weird when you have lots of links because the tooltip triggers for every link as you move it across the sheet.


https://github.com/user-attachments/assets/46abd14e-47b0-4036-8287-3df2cee89946

**Comment by jimniels** [comment]
In tooltips, when we have a keyboard shortcut, we de-emphasize the shortcut itself. Example:

CleanShot 2024-09-20 at 11 23 18@2x

We should do the same here. Something like:

`Open link <span className="opacity-50">(⌘+click)</span>`

**Comment by jimniels** [comment]
**re: Positioning**

> I think the dialogue should be better aligned with the grid, though. Maybe exactly the top left corner of the current cell?

I think the tooltip should work like a tooltip and therefore take its cues for alignment and positioning based on the current position of the user’s cursor. I don't think it should factor in what the grid looks like because, for example, you might be zoomed in or zoomed out and thus the grid's positioning and sizing will change. 

For example, if we align the tooltip on the top left of the cell where the link originates, then the UX is really weird when you have a long link and hover the end of it.

CleanShot 2024-09-23 at 09 40 07

IMO if it looks like a tooltip, it should act like one and therefore appear in proximity to the user's current cursor position regardless of anything else on screen.

---

**Bug: zoom levels**

If you zoom in and hover a link cell, the tooltip is all pixelated and small.

CleanShot 2024-09-23 at 09 43 20@2x

---

**Enhancement: hover interactions**

VSCode provides additional interaction cues when you try to interact with a link. If you hover a link, it gets a subtle background color behind the link text that displays simultaneously with the tooltip, helping bolster the feeling that this element can be interacted with.

Additionally, once you key down on CMD, the link's color changes further bolstering the feeling that this is a link.

CleanShot 2024-09-23 at 09 42 46

I think it'd be nice if we added one (or both) of these.

The background color one could get tricky, especially with cells having the option to have different background colors. So maybe that one we can put on the back burner...

But the text color perhaps we could try a little dash of CSS to help with this effect? I get that people can color their links so we can't depend on them always being blue and thus we can't just set a different shade of blue for the effect.

Perhaps we could use the `filter` property so that when a user hovers a link (or maybe it's on the "cmd" keydown, I'm not entirely sure, play with it) we add a rule along the lines of:

`filter: brightness(1.5)`

Maybe it's a combination of `brightness` and `saturation`, and maybe we don't turn the brightness up but rather turn it down. I'm not sure, I'd say feel free to play with it. But something that adds an additional cue that the link can be interacted with would be a nice touch.

File: quadratic-client/src/app/ui/menus/CommandPalette/commands/Text.tsx [hunk 1]
```diff
@@ -51,6 +61,22 @@ const commands: CommandGroup = {
...
+    },
+    {
+      label: 'Strike-through',
+      isAvailable: isAvailableBecauseCanEditFile,
+      Component: (props) => {
+        return (
+          <CommandPaletteListItem {...props} icon={<FormatStrikethroughIcon />} action={() => setStrikeThrough()} />
+        );
+      },
+    },
```
[the remark this answers is not in the document]
Comment (AyushAgrawal-A2) [reply quadratic-client/src/app/ui/menus/CommandPalette/commands/Text.tsx:79]: not in scope of this PR, command group refactoring is Pending and should be done together

**Review by jimniels — APPROVED** [review]

**Review by davidkircos — APPROVED** [review]

**Pull request merged**